examples/generative/speaker_recognition_using_cnn.py

```diff
@@ -0,0 +1,520 @@
+    num_parallel_calls=tf.data.experimental.AUTOTUNE,
+)
+train_ds = train_ds.prefetch(BATCH_SIZE)
+
+valid_ds = valid_ds.map(
+    lambda x, y: (audio_to_fft(x, None, scale=0), y),
+    num_parallel_calls=tf.data.experimental.AUTOTUNE,
+)
+valid_ds = valid_ds.prefetch(32)
+
+"""
+## Model Definition
+"""
+
+
+def residual_block(x, filters, conv_num=3, activation="relu"):
+    # Shortcut
+    s = keras.layers.Conv1D(filters, 1, padding="same")(x)
+
+    for i in range(conv_num - 1):
+        x = keras.layers.Conv1D(filters, 3, padding="same")(x)
+        x = keras.layers.Activation(activation)(x)
+
+    x = keras.layers.Conv1D(filters, 3, padding="same")(x)
+
+    x = keras.layers.Add()([x, s])
+    x = keras.layers.Activation(activation)(x)
+
+    return keras.layers.MaxPool1D(pool_size=2, strides=2)(x)
+
+
+def build_model(input_shape, num_classes):
+    inputs = keras.layers.Input(shape=input_shape, name="input")
+
+    x = residual_block(inputs, 16, 2)
+    x = residual_block(x, 32, 2)
+    x = residual_block(x, 64, 3)
+    x = residual_block(x, 128, 3)
+    x = residual_block(x, 128, 3)
+
+    x = keras.layers.AveragePooling1D(pool_size=3, strides=3)(x)
+
+    x = keras.layers.Flatten()(x)
+    x = keras.layers.Dense(256, activation="relu")(x)
+    x = keras.layers.Dense(128, activation="relu")(x)
+
+    outputs = keras.layers.Dense(num_classes, activation="softmax", name="output")(x)
+
+    return keras.models.Model(inputs=inputs, outputs=outputs)
+
+
+model = build_model((DESIRED_SR // 2, 1), len(class_names))
+
+model.summary()
+
+# Compile the model using Adam's default learning rate
+model.compile(
+    optimizer="Adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+)
+
+# Let's add callbacks to stop training when the model is not enhancing anymore
+# 'EarlyStopping' and always keep the model that has the best val_accuracy
+# 'ModelCheckPoint'
+earlystopping_cb = keras.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
+mdlcheckpoint_cb = keras.callbacks.ModelCheckpoint(
+    "model.h5", monitor="val_accuracy", save_best_only=True
+)
+
+"""
+## Training
+"""
+
+# Time to train the model
+history = model.fit(
+    train_ds,
+    epochs=EPOCHS,
+    validation_data=valid_ds,
+    callbacks=[earlystopping_cb, mdlcheckpoint_cb],
+)
+
+"""
+## Evaluation
+"""
+
+print(model.evaluate(valid_ds))
+
+"""
+Excellent accuracy!
+"""
+
+"""
+Let's now consider calculating the accuracy when adding noise to the validation set and
+set the noise amplitude to the same magnitude as the audio itself. It will become really
+noisy ...
+"""
+
+valid_ds = audio_dataset_from_directory(
+    DATASET_AUDIO_PATH,
+    desired_channels=1,
+    desired_samples=DESIRED_SR,
+    seed=SHUFFLE_SEED,
+    validation_split=VALID_SPLIT,
+    subset="validation",
+)
+
+valid_ds = valid_ds.map(
+    lambda x, y: (audio_to_fft(x, noises, scale=1.0), y),
+    num_parallel_calls=tf.data.experimental.AUTOTUNE,
+)
+
+print(model.evaluate(valid_ds))
```

Comment:
Please add some form of results visualization (or audition)

Reply:
I took random samples, ran model.predict, displayed the true and predicted speakers along with possibility to listen to the audios